File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/WaitForSpeechToFinish.swift
// ROLE: controller -- `waitForSpeechToFinish`: block until the feed goes quiet.
// BUILT BY: Registry. READS: the session's SpeechBuffer, which owns the heuristic.
// Finished means the buffer stopped growing, not that audio ended: the capture voice receives an
// utterance before any audio exists.

import ScreenReaderWire

public final class WaitForSpeechToFinishHandler: CommandHandler {
	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let params = try request.params(as: WaitToFinishParams.self)
		return WaitToFinishResult(finished: try context.speechBuffer().waitToFinish(timeout: params.timeout))
	}
}

File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/wait_for_speech_to_finish.py
# nvdaMcpBridge domain -- WaitForSpeechToFinishHandler: block until NVDA stops.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: command handler for `waitForSpeechToFinish`; the finish signal is the speech buffer's concern.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from .command_handler import CommandHandler

if TYPE_CHECKING:
	from .session_context import SessionContext


class WaitForSpeechToFinishHandler(CommandHandler):
	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		params = protocol.from_dict(protocol.WaitToFinishParams, request.params)
		return protocol.WaitToFinishResult(finished=ctx.speech_buffer.wait_to_finish(params.timeout))
